File: NapakalakiRuby/pkg/NapakalakiRuby-0.0.1/lib/player.rb
# coding: utf-8

require_relative 'treasure.rb'
require_relative 'bad_consequence.rb'
require_relative 'dice.rb'

module Napakalaki
  
  class Player
    
    # CONSULTORES
    
    def getName
      @name
    end
    
    def getVisibleTreasures
      @visibleTreasures
    end
    
    def getHiddenTreasures
      @hiddenTreasures
    end
    
    def getLevels
      @level
    end
    
    def isDead
      @dead
    end
    
    def canISteal
      @canISteal
    end
    
    def getEnemy
      @enemy
    end
    
    
    # INICIALIZADOR
    
    @@MAXLEVEL=10
    def initialize(name)
      @name=name
      @level=1
      @dead=true
      @canISteal=true
      @visibleTreasures=Array.new
      @hiddenTreasures=Array.new
      @pendingBadConsequence=BadConsequence.newLevelNumberOfTreasures("", 0, 0, 0)
    end
    
    # CONSTRUCTOR DE COPIA
    
    def Player(p)
      @name=p.name
      @level=p.level
      @dead=p.dead
      @canISteal=p.canISteal
      @visibleTreasures=p.visibleTreasures
      @hiddenTreasures=p.hiddenTreasures
      @pendingBadConsequence=p.pendingBadConsequence
      @enemy=p.enemy
    end
    
    # MÉTODOS
    
    
    # Revive al jugador.
    
    private
    def bringToLife
      @dead=false
    end
    
    
    # Calcula el nivel de combate del jugador según las normas del juego.

    def getCombatLevel
      lvl=0
      
      @visibleTreasures.each do |t|
        lvl += t.getBonus
      end
      
      return @level + lvl
    end
    
    
    # Aumenta el nivel del jugador en l niveles.

    def incrementLevels(l)
      @level += l
      
      if(@level > @@MAXLEVEL)
        @level = 10
      end
    end
    
    
    # Decrementa el nivel del jugador en l niveles.

    def decrementLevels(l)
      @level -= l
      
      if(@level <= 0)
        @level = 1
      end
    end
    
    
    # Actualiza el mal rollo por cumplir del jugador.

    def setPendingBadConsequence(b)
      @pendingBadConsequence=b
    end
    
    
    # plica el buen rollo del monstruo m.
    
    def applyPrize(m)
      incrementLevels(m.getLevelsGained)
      
      if(m.getTreasuresGained > 0)
        dealer=CardDealer.instance
        
        i=0
        while(i<m.getTreasuresGained)
          @hiddenTreasures.push(dealer.nextTreasure)
          i = i+1
        end
        
      end
      
    end
    
    
    # Aplica el mal rollo del monstruo m.
    
    def applyBadConsequence(m)
      if(m.getBadConsequence.isDeath)
        discardAllTreasures
      else
        decrementLevels(m.getBadConsequence.getLevels)
        pbc=m.getBadConsequence.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
        setPendingBadConsequence(pbc)
      end

    end
    
    
    # Comprueba si se puede equipar el tesoro t según las reglas del juego.
    
    def canMakeTreasureVisible(t)
      can_make=true
      
      # Sólo se puede tener un tesoro de cada tipo, excepto ONEHAND que se 
      # permite tener 2.
      
      # Recuento de ONEHAND/BOTHHANDS equipados.
      
      one_hand=0
      both_hands=false
      
      @visibleTreasures.each do |x|
        if(x.getType==TreasureKind::ONEHAND)
          one_hand+=1
        end
        if(x.getType==TreasureKind::BOTHHANDS)
          both_hands=true
        end
      end
      if( (one_hand>0 and t.getType==TreasureKind::BOTHHANDS) or
          (both_hands and t.getType==TreasureKind::ONEHAND) )
            can_make=false
      else
        @visibleTreasures.each do |x|
          if( (x.getType==t.getType and x.getType!=TreasureKind::ONEHAND) or
              (t.getType==TreasureKind::ONEHAND and one_hand>1) )
                can_make=false
          end
        end
      end

      return can_make
    end
    
    
    # Devuelve el número de tesoros visibles de tipo tKind que tiene el jugador.

    def howManyVisibleTreasures(tKind)
      total
      
      @visibleTreasures.each do |t|
        if (t.getType == tKind)
          total += 1
        end
      end
      
      return total
    end
    
    
    # Mata al jugador si no tiene tesoros. Implica reiniciar su nivel.
    
    def dieIfNoTreasures
      if (@visibleTreasures.empty? && @hiddenTreasures.empty?)
        @dead = true
        @level = 1      
      end
    end
    
    public
    
    # Desarrolla la lucha del jugador con el monstruo dado. Si el nivel de
    # combate del jugador es mayor que el del monstruo, el jugador gana.
    # Se comprueba si el jugador ha ganado la partida.
    # Devuelve un objeto del tipo CombatResult.
    
    def combat(m)
      
      if(getCombatLevel >= getOponentLevel(m))
        applyPrize(m)
        
        if(@level >= @@MAXLEVEL)
          combat_result=CombatResult::WINGAME
        else
          combat_result=CombatResult::WIN
        end
     
      else
        applyBadConsequence(m)
        if(shouldConvert)
          combat_result=CombatResult::LOSEANDCONVERT
        else
          combat_result=CombatResult::LOSE
        end
        
      end
      
      return combat_result
    end
    
    
    # Equipa el tesoro dado siempre que el jugador pueda hacerlo.
    # Pasa el tesoro t de los tesoros ocultos a los visibles.
    
    def makeTreasureVisible(t)
      
      if(canMakeTreasureVisible(t))
        @visibleTreasures.push(t)
        index=@hiddenTreasures.index(t)
        @hiddenTreasures.delete_at(index)
      end
      
    end
    
    
    # Elimina el tesoro t del los tesoros visibles del jugador. Se elimina 
    # también de su mal rollo pendiente. Por último se comprueba si el jugador
    # se ha quedado sin tesoros, en cuyo caso muere.
    
    def discardVisibleTreasure(t)
      index=@visibleTreasures.index(t)
      @visibleTreasures.delete_at(index)
      
      if(!@pendingBadConsequence.nil? and !@pendingBadConsequence.isEmpty )
            @pendingBadConsequence.substractVisibleTreasure(t)
      end
      
        dieIfNoTreasures();
    end
    
    
    # Elimina el tesoro t del los tesoros ocultos del jugador. Se elimina 
    # también de su mal rollo pendiente. Por último se comprueba si el jugador
    # se ha quedado sin tesoros, en cuyo caso muere.

    def discardHiddenTreasure(t)
      index=@hiddenTreasures.index(t)
      @hiddenTreasures.delete_at(index)
      
      if(!@pendingBadConsequence.nil? and !@pendingBadConsequence.isEmpty )
            @pendingBadConsequence.substractVisibleTreasure(t)
      end
      
        dieIfNoTreasures();
    end
    
    
    # Devuelve true si el jugador no tiene BC que cumplir ni más de 4 tesoros
    # ocultos.
    
    def validState
      valid_state=false
      
      if(@pendingBadConsequence.isEmpty and @hiddenTreasures.count <= 4)
        valid_state=true
      end
      
      return valid_state
    end
    
    
    # Reparte tesoros al jugador en caso de que sea su primer turno o que se
    # haya quedado sin ellos.

    def initTreasures
      dealer=CardDealer.instance
      bringToLife

      @hiddenTreasures.push(dealer.nextTreasure)
      
      dice=Dice.instance
      number=dice.nextNumber
      
      if(number > 1)
        @hiddenTreasures.push(dealer.nextTreasure)
      end
      if(number == 6)
        @hiddenTreasures.push(dealer.nextTreasure)
      end
    end
    
    
    # Roba un tesoro oculto al azar a su enemigo y lo almacena en sus tesoros
    # ocultos. Comprueba que puede robar y que el enemigo tiene tesoros ocultos.
    
    def stealTreasure
      t=nil
      
      if(canISteal)
        if(@enemy.canYouGiveMeATreasure)
          t=@enemy.giveMeATreasure
          @hiddenTreasures.push(t)
          haveStolen
        end
      end
    
      return t
    end
    
    
    # Asigna un enemigo al jugador.
    
    def setEnemy(enemy)
      @enemy=enemy
    end
    
    protected
    # Devuelve un tesoro elegido al azar de entre los tesoros ocultos del jugador.
    
    def giveMeATreasure
      n=Random.rand(@hiddenTreasures.size)
      
      return @hiddenTreasures.delete_at(n)
    end
    
    
    # Devuele true si tiene tesoros(ocultos) que pueden ser robados por otro
    # jugador, falso en caso contrario.

    def canYouGiveMeATreasure
      return !@hiddenTreasures.empty?
    end
    
    
    # Cambia el atributo canISteal a false cuando el jugador roba un tesoro.
    private
    def haveStolen
      @canISteal=false
    end
    
    
    # Descarta todos los tesoros del jugador.
    
    public
    def discardAllTreasures
      visible_treasures_cpy=Array.new(@visibleTreasures)
      hidden_treasures_cpy=Array.new(@hiddenTreasures)
      
      visible_treasures_cpy.each do |t|
        discardVisibleTreasure(t)
      end
      
      hidden_treasures_cpy.each do |t|
        discardHiddenTreasure(t)
      end
      
    end
    
    # Comprueba si el jugador pasa a sectario después de perder un combate.
    
    def shouldConvert
      return Dice.instance.nextNumber==1
    end
    
    
    # Devuelve el nivel de combate del monstruo a combatir.
    
    def getOponentLevel(m)
      m.getCombatLevel
    end
    
    
    # Método to_string
    
    def to_s
      "#{@name}  Nivel: #{@level}  Nivel de combate: #{getCombatLevel}"
    end
  
  end

end

File: Napakalaki/src/napakalaki/Player.java

package napakalaki;

import GUI.Dice;
import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author alex
 */
public class Player {
    
    // DECLARACIÓN DE VARIABLES
    
    private String name;
    private int level;
    private boolean dead;
    private boolean canISteal;
    static int MAXLEVEL = 10;
    
    protected Player enemy;
    private ArrayList<Treasure> visibleTreasures;
    private ArrayList<Treasure> hiddenTreasures;
    private BadConsequence pendingBadConsequence;
    
    
    // CONSTRUCTOR
    
     public Player(String name) {
        this.name = name;
        level = 1;
        dead = true;
        canISteal = true;
        enemy = null;
        visibleTreasures = new ArrayList();
        hiddenTreasures = new ArrayList();
        pendingBadConsequence = new BadConsequence("", false);
    }
     
     
    // CONSTRUCTOR DE COPIA
     
    public Player(Player p){
        name = p.name;
        level = p.level;
        dead = p.dead;
        canISteal = p.canISteal;
        enemy = p.enemy;
        pendingBadConsequence = p.pendingBadConsequence;
        visibleTreasures = new ArrayList(p.visibleTreasures);
        hiddenTreasures = new ArrayList(p.hiddenTreasures);
    }
    
    
    // MÉTODOS

    public Player getEnemy() {
        return enemy;
    }

    public BadConsequence getPendingBadConsequence() {
        return pendingBadConsequence;
    }
    
    public String getName() {
        return name;
    }
    
    
    // Revive al jugador.
    
    private void bringToLife(){
        dead = false;
    }
    
    
    // Calcula el nivel de combate del jugador según las normas del juego.
    
    public int getCombatLevel(){
        int lvl = 0;
        
        for(Treasure t:visibleTreasures)
            lvl += t.getBonus();

        return level + lvl;
    }
    
    
    // Aumenta el nivel del jugador en l niveles.
    
    private void incrementLevels(int l){
        level += l;
        
        if(level > 10)
            level = 10;
    }
    
    
    // Decrementa el nivel del jugador en l niveles.
    
    private void decrementLevels(int l){
        level -= l;
        
        if(level < 1)
            level = 1;
    }
    
    
    // Actualiza el mal rollo por cumplir del jugador.
    
    private void setPendingBadConsequence(BadConsequence b) {
        pendingBadConsequence = b;
    }
    
    
    // Aplica el buen rollo del monstruo m.
    
    private void applyPrize(Monster m){
        incrementLevels(m.getLevelsGained());
        
        if(m.getTreasuresGained() > 0){
            CardDealer dealer = CardDealer.getInstance();
            
            for(int i=0; i<m.getTreasuresGained(); i++)
                hiddenTreasures.add(dealer.nextTreasure());
        }
    }
    
    
    // Aplica el mal rollo del monstruo m.
    
    private void applyBadConsequence(Monster m){
        if(m.getBadConsequence().isDeath())
            discardAllTreasures();
        else{
            decrementLevels(m.getBadConsequence().getLevels());
            BadConsequence pbc = m.getBadConsequence().adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
            setPendingBadConsequence(pbc);
        }
        
    }
    
    
    // Comprueba si se puede equipar el tesoro t según las reglas del juego.
    
    private boolean canMakeTreasureVisible(Treasure t){
        boolean canMake = true;
        
        // Sólo se puede tener un tesoro de cada tipo, excepto ONEHAND que se 
        // permite tener 2.
        
        // 5 tesoros equipados(como máximo) en caso de tener 2 ONEHAND
        // 4 tesoros equipados(como máximo) en caso de tener BOTHHANDS
        
        
        // Recuento de ONEHAND/BOTHHANDS equipados.
        
        int oneHand = 0;
        boolean bothHands = false;
        
        for(Treasure x:visibleTreasures){
            if(x.getType()==TreasureKind.ONEHAND)
                oneHand++;
            if(x.getType()==TreasureKind.BOTHHANDS)
                bothHands=true;
        }
        
        if( (oneHand>0 && t.getType()==TreasureKind.BOTHHANDS) ||
            (bothHands && t.getType()==TreasureKind.ONEHAND) )
                    canMake = false; 
        else
            for(Treasure x:visibleTreasures){
                if( (x.getType()==t.getType() && x.getType() != TreasureKind.ONEHAND)||
                    (t.getType() == TreasureKind.ONEHAND && oneHand>1) )
                        canMake = false;
        }
        
        return canMake;
    }
    
    
    // Devuelve el número de tesoros visibles de tipo tKind que tiene el jugador.
    
    private int howManyVisibleTreasures(TreasureKind tKind){
        int total = 0;
        
        for(Treasure t:visibleTreasures)
            if(t.getType()==tKind)
                total++;
            
        return total;
    }
    
    
    // Mata al jugador si no tiene tesoros. Implica reiniciar su nivel.
    
    private void dieIfNoTreasures(){
        
        if(visibleTreasures.isEmpty() && hiddenTreasures.isEmpty()){
            dead = true;
            level = 1;
        }
            
    }
    
    
    // Devuelve true si el jugador está muerto.
    
    public boolean isDead() {
        return dead;
    }
    
    
    // Devuelve los tesoros visibles del jugador.
    
    public ArrayList<Treasure> getVisibleTreasures() {
        return visibleTreasures;
    }
    
    
    // Devuelve los tesoros ocultos del jugador.

    public ArrayList<Treasure> getHiddenTreasures() {
        return hiddenTreasures;
    }
    
    
    // Desarrolla la lucha del jugador con el monstruo dado. Si el nivel de
    // combate del jugador es mayor que el del monstruo, el jugador gana.
    // Se comprueba si el jugador ha ganado la partida.
    // Devuelve un objeto del tipo CombatResult.
    
    public CombatResult combat(Monster m){
        CombatResult combatResult;
        
        if(getCombatLevel() >= getOponentLevel(m)){
            applyPrize(m);
            
            if(level >= MAXLEVEL)
                combatResult = CombatResult.WINGAME;
            else
                combatResult = CombatResult.WIN;
        }   
        else{
            applyBadConsequence(m);
            if(shouldConvert())
                combatResult = CombatResult.LOSEANDCONVERT;
            else
                combatResult = CombatResult.LOSE;
        }
        
        return combatResult;
    }
    
    
    // Equipa el tesoro dado siempre que el jugador pueda hacerlo.
    // Pasa el tesoro t de los tesoros ocultos a los visibles.
    
    public void makeTreasureVisible(Treasure t){
        if(canMakeTreasureVisible(t)){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }       
    }
    
    
    // Elimina el tesoro t del los tesoros visibles del jugador. Se elimina 
    // también de su mal rollo pendiente. Por último se comprueba si el jugador
    // se ha quedado sin tesoros, en cuyo caso muere.
    
    public void discardVisibleTreasure(Treasure t){
        visibleTreasures.remove(t);
        
        if(pendingBadConsequence != null && !pendingBadConsequence.isEmpty() )
            pendingBadConsequence.substractVisibleTreasure(t);
        
        dieIfNoTreasures();
    }
    
    
    // Elimina el tesoro t del los tesoros ocultos del jugador. Se elimina 
    // también de su mal rollo pendiente. Por último se comprueba si el jugador
    // se ha quedado sin tesoros, en cuyo caso muere.

    public void discardHiddenTreasure(Treasure t){
        hiddenTreasures.remove(t);
        
        if(pendingBadConsequence != null && !pendingBadConsequence.isEmpty() )
            pendingBadConsequence.substractHiddenTreasure(t);
        
        dieIfNoTreasures();
    }
    
    
    // Devuelve true si el jugador no tiene BC que cumplir ni más de 4 tesoros
    // ocultos.
    
    public boolean validState(){
        return (pendingBadConsequence.isEmpty() && hiddenTreasures.size() <= 4);
    }
    
    
    // Reparte tesoros al jugador en caso de que sea su primer turno o que se
    // haya quedado sin ellos.
    
    public void initTreasures(){
        CardDealer dealer = CardDealer.getInstance();
        bringToLife();
        
        hiddenTreasures.add(dealer.nextTreasure());
        
        Dice dice = Dice.getInstance();
        int number = dice.nextNumber();
        
        if(number > 1)
            hiddenTreasures.add(dealer.nextTreasure());
        if(number == 6)
            hiddenTreasures.add(dealer.nextTreasure());
    }
    
    
    // Devuelve el nivel del jugador.
    
    public int getLevel() {
        return level;
    }
    
    
    // Roba un tesoro oculto al azar a su enemigo y lo almacena en sus tesoros
    // ocultos. Comprueba que puede robar y que el enemigo tiene tesoros ocultos.
    
    public Treasure stealTreasure(){
        Treasure t = null;
        
        if(canISteal())
            if(enemy.canYouGiveMeATreasure()){
                t = enemy.giveMeATreasure();
                hiddenTreasures.add(t);
                haveStolen();
            }
                
        return t;
    }
    
    
    // Asigna un enemigo al jugador.
    
    public void setEnemy(Player enemy) {
        this.enemy = enemy;
    }
    
    
    // Devuelve un tesoro elegido al azar de entre los tesoros ocultos del jugador.
    
    protected Treasure giveMeATreasure(){
        Random rnd = new Random();
        int n = rnd.nextInt(hiddenTreasures.size());
        
        return hiddenTreasures.remove(n);
    }
    
    
    // Devuelve true si el jugador puede robar un tesoro.
    
    public boolean canISteal() {
        return canISteal;
    }
    
    
    // Devuelve true si tiene tesoros(ocultos) que pueden ser robados por otro
    // jugador, falso en caso contrario.
    
    protected boolean canYouGiveMeATreasure(){
        return !hiddenTreasures.isEmpty();
    }
    
    
    // Cambia el atributo canISteal a false cuando el jugador roba un tesoro.
    
    private void haveStolen(){
        canISteal = false;
    }
    
    
    // Descarta todos los tesoros del jugador.
    
    public void discardAllTreasures(){
        ArrayList<Treasure> visibleTreasurescpy = new ArrayList(visibleTreasures);
        ArrayList<Treasure> hiddenTreasurescpy = new ArrayList(hiddenTreasures);
        
        for(Treasure t:visibleTreasurescpy)
            discardVisibleTreasure(t);
        for(Treasure t:hiddenTreasurescpy)
            discardHiddenTreasure(t);
    }
    
    
    // Devuelve el nivel del combate del monstruo a combatir.
    
    protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    
    // Devuelve true si cumple las condiciones para convetirse a sectario.
    // Para ello debe perder un combate y sacar un 1 tras lanzar el dado.
    
    protected boolean shouldConvert(){
        Dice dice = Dice.getInstance();
        
        return dice.nextNumber() == 1;
    }
    
    // Método toString

    @Override
    public String toString() {
        if(pendingBadConsequence.isEmpty())
            return name + ", Nivel: " + level + ", Nivel de combate: " + getCombatLevel();
        else
            return name + " \nMAL ROLLO PENDIENTE: " + pendingBadConsequence.toString();            
    }
    
}